x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java

```diff
@@ -570,4 +580,36 @@ private static void checkRemoteEnrich(LogicalPlan plan, Set<Failure> failures) {
             }
         });
     }
+
+    private static void checkFilterMatchConditions(LogicalPlan plan, Set<Failure> failures) {
+        if (plan instanceof Filter f) {
+            Expression condition = f.condition();
+
+            AtomicBoolean hasMatch = new AtomicBoolean(false);
+            condition.forEachDown(MatchQueryPredicate.class, mqp -> {
+               hasMatch.set(true);
+               var field = mqp.field();
+               if (field instanceof FieldAttribute == false) {
+                   failures.add(fail(mqp, "MATCH requires a mapped index field, found [" + field.sourceText() + "]"));
+               }
+
+               if (DataType.isString(field.dataType()) == false) {
+                   var message = LoggerMessageFormat.format(
+                       null,
+                       "MATCH requires a text of keyword field, but [{}] has type [{}]",
+                       field.sourceText(),
+                       field.dataType().esType()
+                   );
+                   failures.add(fail(mqp, message));
+                }
+            });
+
+            if (canPushToSource(condition, x -> false)) {
```

Comment:
`canPushToSource` is a method that's called on data nodes and in certain scenarios it can return `true` or `false` depending on the data mappings that exist on that specific node. I would double check its use here because its usage is meant to be somewhere else.

Reply:
This definitely felt like a code smell - but I think we need to figure out whether a filter using `match` can be pushed down to the lucene query before we apply any of the optimization rules.
Pushing the filters down to the query is done in `LocalPhysicalPlanOptimizer` and I think in the case of `match` we want to fail before the local physical plan stage.

AFAIU the second param of `canPushToSource` (`hasIdenticalDelegate`) looks at whether the field contains a keyword subfield that can be used for exact matching.
Definitely something we need to look closer into, but just returning false for `hasIdenticalDelegate` seemed like the most conservative approach for now.

Reply:
I think everything we do at analysis time generally is done on data type - and maybe the indexed flag. But if something is of type `text` then we can do a `MATCH`, right? If it doesn't have a search index (which would be weird) we *can* analyze on the fly. At least, eventually we can. For now we could just fail?

Reply:
I talked to @ioanatia about this. It's to fail nicely with stuff like `|WHERE title MATCH "harry potter" OR SIGNUM(year) == 1`. This can't be pushed and we don't have an implementation of `MATCH` that can run at compute engine time. This is the first thing like that. Without this test we get a rather exciting error message somewhere else. I'd prefer not to do that.

Two important things:
1. @ioanatia had a wonderful way to run the `MATCH` at compute engine time - unpushed. Sort of via a rescore-like mechanism. I think we can do that, but it's it's own thing. I'm not sure it's high on her priority list, though it does feel important to me. Just to make ESQL stay super composable.
2. If we have a better way to make this error message I'm happy with it. But this *is* temporary. We will figure out a way to run without pushing.

Reply:
I added a comment explaining what we are doing here - with that in mind I am solving the comment since this is just a temporary state.